src/Controller/BackendCsvImportController.php

```diff
@@ -0,0 +1,356 @@
+    }
+
+    /**
+     * Returns an array of data from imported CSV files.
+     *
+     * @param FileUpload $uploader
+     * @param string     $separator
+     * @param callable   $callback
+     *
+     * @return array
+     */
+    private function fetchData(FileUpload $uploader, $separator, callable $callback)
+    {
+        $data      = [];
+        $files     = $this->getFiles($uploader);
+        $delimiter = $this->getDelimiter($separator);
+
+        foreach ($files as $file) {
+            $fp = fopen($file, 'rb');
+
+            while (($row = fgetcsv($fp, 0, $delimiter)) !== false) {
+                $data = $callback($data, $row);
+            }
+        }
+
+        return $data;
+    }
+
+    /**
+     * Returns the form ID for the template.
+     *
+     * @param Request $request
+     *
+     * @return string
+     */
+    private function getFormId(Request $request)
+    {
+        return 'tl_csv_import_'.$request->query->get('key');
+    }
+
+    /**
+     * Returns the back button and redirect URL.
+     *
+     * @param Request $request
+     *
+     * @return string
+     */
+    private function getBackUrl(Request $request)
+    {
+        return str_replace('&key='.$request->query->get('key'), '', $request->getRequestUri());
+    }
+
+    /**
+     * Returns an array of separators for the template.
+     *
+     * @param bool $allowLinebreak
+     *
+     * @return array
+     */
+    private function getSeparators($allowLinebreak = false)
+    {
+        $separators = [
+            self::SEPARATOR_COMMA     => [
+                'delimiter' => ',',
+                'value'     => self::SEPARATOR_COMMA,
+                'label'     => $GLOBALS['TL_LANG']['MSC']['comma'],
+            ],
+            self::SEPARATOR_SEMICOLON => [
+                'delimiter' => ';',
+                'value'     => self::SEPARATOR_SEMICOLON,
+                'label'     => $GLOBALS['TL_LANG']['MSC']['semicolon'],
+            ],
+            self::SEPARATOR_TABULATOR => [
+                'delimiter' => "\t",
+                'value'     => self::SEPARATOR_TABULATOR,
+                'label'     => $GLOBALS['TL_LANG']['MSC']['tabulator'],
+            ],
+        ];
+
+        if ($allowLinebreak) {
+            $separators[self::SEPARATOR_LINEBREAK] = [
+                'delimiter' => "\n",
+                'value'     => self::SEPARATOR_LINEBREAK,
+                'label'     => $GLOBALS['TL_LANG']['MSC']['linebreak'],
+            ];
+        }
+
+        return $separators;
+    }
+
+    /**
+     * Converts a separator name/constant into a delimiter character.
+     *
+     * @param string $separator
+     *
+     * @return string
+     *
+     * @throws \RuntimeException
+     */
+    private function getDelimiter($separator)
+    {
+        $separators = $this->getSeparators(true);
+
+        if (!isset($separators[$separator])) {
+            throw new \RuntimeException($GLOBALS['TL_LANG']['MSC']['separator'][1]);
```

Comment:
🤔 two issues with that:
 1. `\InvalidArgumentException` is a `LogicException`, which is not really correct for user input.
 2. We're currently using `RuntimeException` to add flashbag messages, so that would not work anymore…